server/controllers/refer/reasonController.test.ts

```diff
@@ -0,0 +1,113 @@
+  const referralService = createMock<ReferralService>({})
+
+  const courseOffering = courseOfferingFactory.build()
+
+  let reasonController: ReasonController
+
+  beforeEach(() => {
+    request = createMock<Request>({ user: { token } })
+    response = createMock<Response>({})
+    reasonController = new ReasonController(personService, referralService)
+  })
+
+  afterEach(() => {
+    jest.resetAllMocks()
+  })
+
+  describe('show', () => {
+    it('renders the reason for referral page', async () => {
+      const person = personFactory.build()
+      personService.getPerson.mockResolvedValue(person)
+
+      const referral = referralFactory
+        .started()
+        .build({ offeringId: courseOffering.id, prisonNumber: person.prisonNumber })
+      referralService.getReferral.mockResolvedValue(referral)
+
+      const emptyErrorsLocal = { list: [], messages: {} }
+      ;(FormUtils.setFieldErrors as jest.Mock).mockImplementation((_request, _response, _fields) => {
+        response.locals.errors = emptyErrorsLocal
+      })
+
+      const requestHandler = reasonController.show()
+      await requestHandler(request, response, next)
+
+      expect(response.render).toHaveBeenCalledWith('referrals/reason', {
+        pageHeading: 'Add reason for referral and any additional information',
+        person,
+        referral,
+      })
+
+      expect(FormUtils.setFieldErrors).toHaveBeenCalledWith(request, response, ['reason'])
+    })
+
+    describe('when the person service returns `null`', () => {
+      it('responds with a 404', async () => {
+        const person = personFactory.build()
+        personService.getPerson.mockResolvedValue(null)
+
+        const referral = referralFactory
+          .started()
+          .build({ offeringId: courseOffering.id, prisonNumber: person.prisonNumber })
+        referralService.getReferral.mockResolvedValue(referral)
+
+        const requestHandler = reasonController.show()
+        const expectedError = createError(404, `Person with prison number ${referral.prisonNumber} not found.`)
+
+        expect(() => requestHandler(request, response, next)).rejects.toThrowError(expectedError)
+      })
+    })
+  })
+
+  describe('update', () => {
+    it('updates the referral and redirects to the referral show page', async () => {
+      const referral = referralFactory.started().build()
+      referralService.getReferral.mockResolvedValue(referral)
+
+      request.body.reason = ' Some reason\nAnother paragraph\n '
+
+      const requestHandler = reasonController.update()
+      await requestHandler(request, response, next)
+
+      expect(referralService.updateReferral).toHaveBeenCalledWith(token, referral.id, {
+        oasysConfirmed: referral.oasysConfirmed,
+        reason: 'Some reason\nAnother paragraph',
+      })
+      expect(response.redirect).toHaveBeenCalledWith(`/referrals/${referral.id}`)
+    })
+
+    describe('when the reason is not provided', () => {
```

Comment:
Might be good to also test providing a reason that's just space characters (`  ` or ` \n  `) - not sure if it's worth integration testing this too, but probably not?